### DISCUSSION TITLE
Custom Level and Properties does not work together

### DISCUSSION THREAD

**Discussion by paulonicolau-rj** [apache/logging-log4j2#2279] [Q&A]
## Description

I´m implementing a log file contains trace of request for a funcionality. When i use absolute path in filename attribute of RoliingFileappender, it works fine. When I use property to declare filename, this is ignored and his key is used. 

## Configuration

**Version:** [Log4j 2.22.1]

**Operating system:** [Windows 10/Amazon Linux 2]

**JDK:** [Oracle 17]

## Logs

```
[Stacktraces, errors, etc. relevant applications logs.]
```

## Reproduction

[An isolated test reproducing the test.
JUnit tests similar to the ones in the code base are extremely appreciated.]

![image](https://github.com/apache/logging-log4j2/assets/78874860/495e3de7-bf28-46ff-99ab-762835498fc8)
![image](https://github.com/apache/logging-log4j2/assets/78874860/ce912784-3481-4ec7-8e98-f934818e7288)

**Comment by RapidTransit** [comment]
Version: 2.21.1
I had a similar problem, I don't recall what fixed it but here's a snippet of my config, I want to say maybe the property was coming up null?:
AFAIK I haven't tweaked my production config.
```   
 <Properties>
        <Property name="log4j2.disableJmx">true</Property>
        <Property name="log4j2.disable.jmx">true</Property>
        <Property name="log4j2.enableJndiLookup">false</Property>
        <Property name="filename">${env:PROXY_LOG_LOCATION:-/var/log/proxy}</Property>
    </Properties>

    <Select>
    <SystemPropertyArbiter propertyName="log.environment" propertyValue="production">

        <Appenders>

            <RollingFile name="GenericRollingFile" fileName="${filename}/application.log" filePattern="logs/application-%d{MM-dd-yyyy}-%i.log.gz">
```

**Comment by paulonicolau-rj** [comment]
I can´t see, but in my case, the problem is when i use a customlevels. If I remove the tag CustomLevels, all works fine.

**Comment by ppkarwasz** [comment]
@paulonicolau-rj,

Can you add your full configuration as text? Since all other nodes are interpolated using the properties that you provide, the `<Properties>` element must be the **first** child of `<Configuration>`.

**Comment by paulonicolau-rj** [comment]
> @paulonicolau-rj,
> 
> Can you add your full configuration as text? Since all other nodes are interpolated using the properties that you provide, the `<Properties>` element must be the **first** child of `<Configuration>`.

Yes, follow:
<code>
    <?xml version="1.0" encoding="UTF-8"?>
    
    <Configuration status="warn">
	    <CustomLevels>
		    <CustomLevel name="EVENT" intLevel="550" />
	    </CustomLevels>
    
	    <Properties>
		    <Property name="basePath">${sys:catalina.home}/logs</Property>
	    </Properties>
    
	    <Appenders>
		    <RollingFile name="eventFile" fileName="${basePath}/api-event.log" filePattern="${basePath}/api-event-%d{yyyy-MM-dd}.log">
			    <LevelRangeFilter minLevel="EVENT" maxLevel="EVENT" onMatch="ACCEPT" onMismatch="DENY"/>
			    <PatternLayout>
				    <pattern>[%-5level] %d{yyyy-MM-dd HH:mm:ss.SSS} [%t] %c{1} - %msg%n</pattern>
			    </PatternLayout>
			    <Policies>
				    <TimeBasedTriggeringPolicy interval="1" modulate="true" />
				    <SizeBasedTriggeringPolicy size="10MB" />
			    </Policies>
			    
			    <DefaultRolloverStrategy max="10">
				    <Delete basePath="${basePathr}" maxDepth="10">
					    
					    <IfLastModified age="30d" />
				    </Delete>
			    </DefaultRolloverStrategy>
		    </RollingFile>
		    
		    <RollingFile name="infoFile" fileName="${basePath}/api-info.log" filePattern="${basePath}/api-info-%d{yyyy-MM-dd}.log">
			    <LevelRangeFilter minLevel="INFO" maxLevel="INFO" onMatch="ACCEPT" onMismatch="DENY"/>
			    <PatternLayout>
				    <pattern>[%-5level] %d{yyyy-MM-dd HH:mm:ss.SSS} [%t] %c{1} - %msg%n</pattern>
			    </PatternLayout>
			    <Policies>
				    <TimeBasedTriggeringPolicy interval="1" modulate="true" />
				    <SizeBasedTriggeringPolicy size="10MB" />
			    </Policies>
			    
			    <DefaultRolloverStrategy max="10">
				    <Delete basePath="${basePathr}" maxDepth="10">
					    
					    <IfLastModified age="30d" />
				    </Delete>
			    </DefaultRolloverStrategy>
		    </RollingFile>
    
		    <RollingFile name="errorFile" fileName="${basePath}/api-error.log" filePattern="${basePath}/api-error-%d{yyyy-MM-dd}.log">
			    <LevelRangeFilter minLevel="ERROR" maxLevel="ERROR" onMatch="ACCEPT" onMismatch="DENY"/>
			    <PatternLayout>
				    <pattern>[%-5level] %d{yyyy-MM-dd HH:mm:ss.SSS} [%t] %c{1} - %msg%n</pattern>
			    </PatternLayout>
			    <Policies>
				    <TimeBasedTriggeringPolicy interval="1" modulate="true" />
				    <SizeBasedTriggeringPolicy size="10MB" />
			    </Policies>
			    
			    <DefaultRolloverStrategy max="10">
				    <Delete basePath="${basePathr}" maxDepth="10">
					    
					    <IfLastModified age="30d" />
				    </Delete>
			    </DefaultRolloverStrategy>
		    </RollingFile>
	    </Appenders>
    
	    <Loggers>
		    <Root level="all" additivity="false">
			    <appender-ref ref="eventFile" level="event"/>
			    <appender-ref ref="infoFile" level="info"/>
			    <appender-ref ref="errorFile" level="error" />
		    </Root>
	    </Loggers>
    </Configuration>
</code>

**Comment by ppkarwasz** [comment]
Switch the order of `<Properties>` and `<CustomLevels>`:

```xml
<Configuration>
    <Properties>
	    <Property name="basePath">${sys:catalina.home}/logs</Property>
    </Properties>

    <CustomLevels>
	    <CustomLevel name="EVENT" intLevel="550" />
    </CustomLevels>

    ...
</Configuration>
```

**Comment by paulonicolau-rj** [comment]
> Switch the order of `<Properties>` and `<CustomLevels>`:
> 
> ```
> <Configuration>
>     <Properties>
> 	    <Property name="basePath">${sys:catalina.home}/logs</Property>
>     </Properties>
> 
>     <CustomLevels>
> 	    <CustomLevel name="EVENT" intLevel="550" />
>     </CustomLevels>
> 
>     ...
> </Configuration>
> ```

Thanks, it worked partially. Now, it´s working when I use absolute path like a value. 
`<Property name="basePath">/usr/share/tomcat/logs</Property>`

But, if I use a system parameter, por example, it doesn´t work.
`<Property name="basePath">${sys:catalina.home}/logs</Property>`

**Comment by ppkarwasz** [comment]
Use:

```xml
<Property name="basePath" value="${sys:catalina.home}/logs"/>
```

<s>Unlike XML attributes, the content of the `<Property>` element is not interpolated. This might actually be a bug:</s>

https://github.com/apache/logging-log4j2/blob/397771c0feb56e00482fd6230e1e124b7c483333/log4j-core/src/main/java/org/apache/logging/log4j/core/config/Property.java#L134-L137

**Comment by paulonicolau-rj** [comment]
So, it worked fully. But, I consider the first report a bug too. 
Thanks friend!

**Reply by ppkarwasz** [reply]
FYI: The restriction on the order of the `<Properties>` element was lifted in 2.23.0 (cf. #2278).

**Comment by ppkarwasz** [comment]
@paulonicolau-rj,

When you add other elements before `<Properties>` you get a ""Properties declaration must be the first element in the configuration" error message on your stdout.

**Comment by paulonicolau-rj** [comment]
I can´t see this message in stdout, but what I can see is the System property catalina.home is initilize before the container and so, it is store the log files in another folder of eclipse workspace. But this there isn´t a problem.

**Comment by ppkarwasz** [comment]
Regarding the second problem you reported, I am not able to reproduce it. **Both**:

```xml
<Property name="basePath" value="${sys:catalina.home}/logs"/>
```

and 

```xml
<Property name="basePath">${sys:catalina.home}/logs</Property>
```

work for me.
